### PR TITLE
Support caching

### LINKED ISSUE #29: Support for caching SQL-ingested data
This should be relatively easy. When a `-c`, `--cache` flag is on, instead of [generating a random project name](https://github.com/multiprocessio/dsq/blob/main/main.go#L163), generate one based on the hashed contents of the files being queried.

Don't use an [in-memory :memory: SQLite table](https://github.com/multiprocessio/dsq/blob/main/main.go#L275) but put the SQLite file on disk with the same name as the project (based on the hash).

But in actually eval-ling the SQL/Program panel there might need to be a new flag added to DataStation and set in the evalContext maybe that says "don't reingest" and just run the SQL query with the DM_getPanel() calls substituted as usual for table names in SQL.

Finally, just [don't delete the generated directory at the end of the script](https://github.com/multiprocessio/dsq/blob/main/main.go#L167).

### PR DESCRIPTION
**Pull request by Posrabi**
- Add "--cache"" flag
- Create sqlite db file on disk with name as the hash of its content.
- Add test entry.
* This will fail tests until the changes in datastation is committed.

Closes #29

### CODE REVIEW

**Comment by Posrabi** [comment]
initial_cache
after_cache
//user-images.githubusercontent.com/87549865/163600422-e0e0e5c8-0013-4fe2-824e-c01048dd667c.svg)

File: main.go [hunk 1]
```diff
@@ -162,6 +164,30 @@ func dumpJSONFile(file string, pretty bool, schema bool) error {
...
+			return "", err
+		}
+		contents = append(contents, c...)
+	}
+
+	sha1 := sha1.New()
+	_, err := sha1.Write(contents)
+	if err != nil {
+		return "", err
+	}
```
Comment (eatonphil) [inline main.go:182]: I think we can reorganize this a bit to not put the whole thing in memory all the time:

```
sha1 := sha1.New()
for _, f := range files {
  f, err := os.Open()
  // handle err

  _, err = io.Copy(f, sha1)
  // handle err
}

return hex.EncodeToString(sha1.Sum(nil)), nil
```

File: main.go [hunk 2]
```diff
@@ -164,21 +164,50 @@ func dumpJSONFile(file string, pretty bool, schema bool) error {
...
-		c, err := os.ReadFile(f)
-		if err != nil {
-			return "", err
-		}
-		contents = append(contents, c...)
-	}
+	for i := 0; i < len(files); i++ {
+		if files[i] == "" {
+			i += 1
+			b, err := io.ReadAll(os.Stdin)
```
Comment (eatonphil) [inline main.go:173]: I'm still confused why you're doing temp buffers here.

You should be able to hook up os.Stdin directly to sha1.Writer without a temp buffer.

Comment (eatonphil) [reply main.go:173]: I mentioned io.Copy before, here's another example of that https://stackoverflow.com/a/45123990/1507139

Comment (eatonphil) [reply main.go:173]: Which is pulled from the sha1 docs https://pkg.go.dev/crypto/sha256#New

Comment (eatonphil) [reply main.go:173]: Although this isn't a file since os.Stdin is a writer I think it will work the same way

File: main.go [hunk 3]
```diff
@@ -162,6 +164,50 @@ func dumpJSONFile(file string, pretty bool, schema bool) error {
...
+			if _, err := file.Write(b); err != nil {
+				return "", err
+			}
+
+		} else {
+			file, err := os.Open(files[i])
+			if err != nil {
+				return "", err
+			}
+			defer file.Close()
```
Comment (eatonphil) [inline main.go:196]: defer happens at the end of functions only, not blocks. So you can't defer you need to just call `file.close` at the end of the loop

File: main.go [hunk 4]
```diff
@@ -162,6 +164,50 @@ func dumpJSONFile(file string, pretty bool, schema bool) error {
...
+			b, err := io.ReadAll(os.Stdin)
+			if err != nil {
+				return "", err
+			}
+
+			if _, err := sha1.Write(b); err != nil {
+				return "", err
+			}
+
+			file, err := os.OpenFile(tmp, os.O_RDWR, os.ModePerm)
```
Comment (eatonphil) [inline main.go:182]: You need to close this too I think?

File: main.go (hunk 3 above)
Comment (eatonphil) [reply main.go:196]: btw file closing doesn't matter on anywhere except Windows. If one process has a file open no one else can edit it (or read it I can't remember)

File: main.go [hunk 5]
```diff
@@ -269,12 +329,65 @@ func _main() error {
...
+	ec := runner.NewEvalContext(settings, tmpDir)
+
+	if cache {
+		// sqlite file is present on disk
+		if info, err := os.Stat(getCachedDBPath(project.Id)); err != nil || info.Size() == 0 {
+			file, err := openTruncate(getCachedDBPath(project.Id))
+			if err != nil {
+				return err
+			}
+			file.Close()
```
Comment (eatonphil) [inline main.go:346]: Why are you opening it just to close it?

File: main.go [hunk 6]
```diff
@@ -269,12 +329,65 @@ func _main() error {
...
 	defer os.RemoveAll(tmpDir)
+	if cache {
+		tmpDir = "dsq-cache"
+	}
+
+	settings := *runner.DefaultSettings
+	ec := runner.NewEvalContext(settings, tmpDir)
+
+	if cache {
+		// sqlite file is present on disk
```
Comment (eatonphil) [inline main.go:340]: I would move all of this into a separate function that returns an error.

If the error is not nil we should not fail we should fall back to not using the cached version. So it should just log the error and keep going

File: main.go [hunk 7]
```diff
@@ -297,6 +410,13 @@ func _main() error {
 				}
 				i += 1
 
+				// stdin has been exhausted after hashing
+				if cache {
+					b, err = io.ReadAll(projectTmp)
```
Comment (eatonphil) [inline main.go:415]: ReadAll is a bad sign. Why are you doing this?

File: main.go [hunk 8]
```diff
@@ -391,6 +518,9 @@ func _main() error {
 	}
 
 	resultFile := ec.GetPanelResultsFile(project.Id, panel.Id)
+	if cache {
+		defer os.Remove(resultFile)
```
Comment (eatonphil) [inline main.go:522]: You can just remove this in all cases even if it's redundant. But you should leave a comment as to why it's necessary

File: scripts/test.py [hunk 9]
```diff
@@ -213,6 +213,32 @@ def test(name, to_run, want, fail=False, sort=False):
 
 # END OF REGRESSION TESTS
 
+# Cache test
+to_run = """
+curl https://s3.amazonaws.com/nyc-tlc/trip+data/yellow_tripdata_2021-04.csv -o taxi.csv;
+./dsq --cache taxi.csv "SELECT passenger_count, COUNT(*), AVG(total_amount) FROM {} GROUP BY passenger_count";
```
Comment (eatonphil) [inline scripts/test.py:219]: Although the original script doesn't include it you should add an `ORDER BY  COUNT(*) DESC` otherwise by the SQL spec the order could be random. I don't know if SQLite picks an order automatically or if it's really allowed to be random.

File: main.go (hunk 7 above)
Comment (eatonphil) [reply main.go:415]: I think it would be better to edit like 421 and and wherever `b` is first defined have it also do:

```
var r io.Reader = bytes.NewReader(b)
```

so that here you can do something like 

```
if cache {
  r, err = io.Open(projectTmp)
}
```

And not have to read the whole thing twice

File: main.go [hunk 10]
```diff
@@ -269,12 +329,65 @@ func _main() error {
...
+				return err
+			}
+			file.Close()
+		} else {
+			settings.CacheMode = true
+			ec = runner.NewEvalContext(settings, tmpDir)
+		}
+
+		notPresentFiles := []string{}
+		for i := 0; i < len(files); i++ {
```
Comment (eatonphil) [inline main.go:353]: I am pretty sure there's a way to do this without duplicating all of this code. Something like just adding the file name read from stdin to the list of `files`.

File: main.go (hunk 2 above)
Comment (eatonphil) [reply main.go:173]: btw just one

File: main.go (hunk 7 above)
Comment (Posrabi) [reply main.go:415]: Because the existing code is also ReadAll from os.Stdin, the projectTmp file is just a copy of os.Stdin in this case. So should I remove both  readAlls?

File: main.go [hunk 11]
```diff
@@ -254,9 +306,17 @@ func _main() error {
 	if err != nil {
 		return err
 	}
+	projectID := projectTmp.Name()
+	if cache {
+		projectID, err = getFilesContentHash(files, projectTmp.Name())
```
Comment (eatonphil) [inline main.go:311]: You could have this getFilesContentHash return another variable that indiciates if it copied from stdin to a tmp file.

That way you don't need that second loop through `files` without all the duplicated code you added.

It could be something like:

```
projectId, stdinTmpCopy, err = getFilesContentHash(files, projectTmp.Name())
if stdinTmpCopy != "" {
  files = append(files, stdinTmpCopy)
}
```

File: main.go (hunk 6 above)
Comment (eatonphil) [reply main.go:340]: After my other comments I think you just don't need this section at all.

But still, if anything fails in cache mode it shouldn't fail the whole command it should just log and fall back to the non-cached version.

File: main.go (hunk 7 above)
Comment (eatonphil) [reply main.go:415]: Take a look at my second comment and see if that makes sense

File: main.go (hunk 5 above)
Comment (Posrabi) [reply main.go:346]: I am creating the sqlite file.

Comment (eatonphil) [reply main.go:346]: Why does that need to happen?

File: main.go (hunk 2 above)
Comment (Posrabi) [reply main.go:173]: Oh I  just didn't see that comment prior to the first commit

File: main.go [hunk 12]
```diff
@@ -254,30 +309,60 @@ func _main() error {
 	if err != nil {
 		return err
 	}
+	projectID := projectTmp.Name()
+	if cache {
+		stdinTmpCopy := false
+		projectID, stdinTmpCopy, err = getFilesContentHash(files, projectTmp.Name())
+		if err != nil {
+			fmt.Printf("Error creating hash for cache mode: %v, defaulting to normal mode", err)
```
Comment (eatonphil) [inline main.go:317]: Everywhere you're using fmt.Println you should use log.Println instead. This is because fmt.Println goes to stdout and log.Println goes to stderr. So if someone does shell redirection `dsq blah blah > x.json` fmt.Println stuff goes to x.json whereas log.Println stuff doesn't unless you do `dsq blah blah 2&>1 > x.json`.

File: main.go [hunk 13]
```diff
@@ -254,30 +309,60 @@ func _main() error {
...
 	}
 	defer os.RemoveAll(tmpDir)
+	if cache {
+		tmpDir = "dsq-cache"
+		settings.CacheMode = &cache
+	}
+	ec := runner.NewEvalContext(settings, tmpDir)
 
-	ec := runner.NewEvalContext(*runner.DefaultSettings, tmpDir)
 	for i := 0; i < len(files); i++ {
```
Comment (eatonphil) [inline main.go:347]: I think that when we're in cache mode and the cache is valid, this loop shouldn't run at all.

The only reason we run it is to grab shapes because the evaldatabase call below expects the project to have valid shapes.

But when we've already cached we don't need to check shapes. Shapes are only used during imports, not during query.

So I think we should find a way to skip this loop completely when the cache is valid.

Comment (Posrabi) [reply main.go:347]: hmm okay, that means we don't have to save all the json files too.

Comment (eatonphil) [reply main.go:347]: Right! The only thing that needs to stick around is the sqlite file.

File: main.go [hunk 14]
```diff
@@ -254,95 +310,131 @@ func _main() error {
...
+	if cache == nil || *cache == false {
+		for i := 0; i < len(files); i++ {
+			file := files[i]
+			panelId := uuid.New().String()
+			resultFile := ec.GetPanelResultsFile(project.Id, panelId)
+			out, err := openTruncate(resultFile)
+			if err != nil {
+				return err
+			}
+			defer os.Remove(resultFile)
```
Comment (eatonphil) [inline main.go:362]: You can't defer in a loop

File: main.go [hunk 15]
```diff
@@ -254,95 +310,131 @@ func _main() error {
...
-				mimetype := files[i+1]
-				if !strings.Contains(mimetype, string(filepath.Separator)) {
-					mimetype = string(runner.GetMimeType("x."+mimetype, runner.ContentTypeInfo{}))
-				}
-
-				if mimetype == "" {
-					return fmt.Errorf("Unknown mimetype or file extension: %s.", mimetype)
+	ec := runner.NewEvalContext(settings, tmpDir)
+	if cache == nil || *cache == false {
+		for i := 0; i < len(files); i++ {
```
Comment (eatonphil) [inline main.go:354]: Getting pretty nested inside here. Let's move this into a new function?

```
if cache == nil || *cache == false {
  importFiles(files)
}
```

Comment (Posrabi) [reply main.go:354]: sure!

File: main.go [hunk 16]
```diff
@@ -334,106 +432,26 @@ func _main() error {
...
-					Shape: *s,
-				},
-				Id:   panelId,
-				Name: uuid.New().String(),
-			})
-
-			out.Close()
+	ec := runner.NewEvalContext(*runner.DefaultSettings, tmpDir)
+	toDelete := []string{}
+	if !cacheSettings.CachePresent || !cacheSettings.Enabled {
```
Comment (eatonphil) [inline main.go:451]: This looks more readable, thank you!

File: main.go [hunk 17]
```diff
@@ -491,8 +509,11 @@ func _main() error {
 	}
 
 	resultFile := ec.GetPanelResultsFile(project.Id, panel.Id)
-	defer os.Remove(resultFile) // in cache mode, dsq-cache dir is not remove.
-	return dumpJSONFile(resultFile, pretty, schema)
```
Comment (eatonphil) [inline main.go:394]: To simplify this slightly you can:

```
toDelete = append(toDelete, resultFile)
defer cleanup(toDelete)
return dumpJSONFile(resultFile, pretty, schema)
```

File: main.go [hunk 18]
```diff
@@ -239,14 +239,113 @@ Examples:
...
+		out.Close()
+	}
+
+	return toDelete, nil
+}
+
+func cleanUp(files []string) error {
+	for _, f := range files {
+		if err := os.Remove(f); err != nil {
+			return err
```
Comment (eatonphil) [inline main.go:335]: Let's have it just log and continue instead of failing

File: main.go [hunk 19]
```diff
@@ -162,6 +164,53 @@ func dumpJSONFile(file string, pretty bool, schema bool) error {
...
+		} else {
+			file, err := os.Open(files[i])
+			if err != nil {
+				return "", false, err
+			}
+			_, err = io.Copy(sha1, file)
+			if err != nil {
+				return "", false, err
+			}
+			file.Close()
```
Comment (eatonphil) [inline main.go:203]: Ok the last thing that looks sketchy is that that the close code here and in three other places in this PR is so far from the open.

Now I know I said you can't use defer in a loop. But these loop bodies are big anyway.

So what I do in these situations is break out the body into its own function so that I *can* call defer. That way the close code is nearby the open code.

When important cleanup code is far from the initialization code you're just asking for bugs.

I'm totally ok with merging this PR wihtout this and I can do this cleanup myself because this PR has dragged on so long and I feel bad :D Or you can do it here. Either way is ok! But it should be all 3 places: line 203, line 191 and line 326

Comment (Posrabi) [reply main.go:203]: I will clean it up, it's only a couple more lines.

File: main.go [hunk 20]
```diff
@@ -162,6 +164,62 @@ func dumpJSONFile(file string, pretty bool, schema bool) error {
 	return nil
 }
 
+func getFilesContentHash(files []string, tmp string) (string, bool, error) {
+	sha1 := sha1.New()
```
Comment (eatonphil) [inline main.go:168]: Do you want to sort the files here?

Comment (eatonphil) [reply main.go:168]: But you'd need to sort a copy of this files list because order is important in the query.

File: main.go [hunk 21]
```diff
@@ -162,6 +164,62 @@ func dumpJSONFile(file string, pretty bool, schema bool) error {
...
+	stdinTmpCopy := false
+
+	for i := 0; i < len(files); i++ {
+		err := func() error {
```
Comment (eatonphil) [inline main.go:172]: It doesn't really help code *structure* to use a nested function literal like this. I'd normally just break it out into a separate top-level function. That normally also makes it easier if you need to unit test it.

Comment (eatonphil) [reply main.go:172]: But you don't have to make that change right now. It's ok for now

File: main.go (hunk 20 above)
Comment (Posrabi) [reply main.go:168]: Uhm we would also to sort the table order in the query as well. Because the order in the database is already set. Like if our initial call is 
./dsq --cache  testdata/join/ages.json testdata/join/users.csv  "select u.name, a.age from {1} u join {0} a on u.id = a.id"
Our second call of:
./dsq --cache testdata/join/users.csv  testdata/join/ages.json   "select u.name, a.age from {0} u join {1} a on u.id = a.id"
wouldn't work.

Comment (Posrabi) [reply main.go:168]: Would sorting that too has unintended side effects?

Comment (Posrabi) [reply main.go:168]: I am testing it with the copy, the order in the query wouldn't be correct because the order in the sqlite file is according to the first query

Comment (eatonphil) [reply main.go:168]: Oh yeah that's a good point. OK! No need for sorting

File: main.go [hunk 22]
```diff
@@ -391,6 +516,8 @@ func _main() error {
 	}
 
 	resultFile := ec.GetPanelResultsFile(project.Id, panel.Id)
+	toDelete = append(toDelete, resultFile)
```
[the remark this answers is not in the document]
Comment (eatonphil) [reply main.go:519]: You resolved this but didn't comment and didn't change it. Any thoughts?

[the remark this answers is not in the document]
Comment (Posrabi) [reply main.go]: I did remove the toDelete slice.

[the remark this answers is not in the document]
Comment (Posrabi) [reply main.go]: Just pushed the latest commit.

File: main.go [hunk 23]
```diff
@@ -190,13 +252,106 @@ Examples:
...
+		Name: uuid.New().String(),
+	})
+
+	return i, nil
+}
+
+func importFiles(cacheMode bool, files []string, ec *runner.EvalContext, project *runner.ProjectState) error {
+	var err error
+	for i := 0; i < len(files); i++ {
+		if i, err = importFilesIteration(cacheMode, &files, ec, project, i); err != nil {
```
Comment (eatonphil) [inline main.go:340]: Why do you need to pass in files and i separately? Why can't you just pass in `files[i]`?

Also, you can just call this `importFile` rather than `importFilesIteration` unless there's a reason not to use `importFile`.

File: main.go [hunk 24]
```diff
@@ -190,13 +252,106 @@ Examples:
...
+			var r io.Reader = bytes.NewReader(b)
+			if cacheMode {
+				r, err = os.Open(files[len(files)-1])
+				if err != nil {
+					return i, fmt.Errorf("Error opening copied stdin file: %v", err) // returns as it's not possible to go back to default.
+				} else {
+					*filesPtr = files[:len(files)-1]
+				}
+			}
+			fmt.Println(files)
```
Comment (eatonphil) [inline main.go:291]: Reminder to `git grep fmt.Println` and delete this before merging

File: main.go (hunk 23 above)
Comment (eatonphil) [reply main.go:340]: Yeah I see how you depend on `i` inside this function but I don't think that's a good idea.

I'm really sorry about dragging out this PR though we can merge it and I can do some followup cleanup.

I don't want you to keep having to edit stuff.

Comment (Posrabi) [reply main.go:340]: We need to modify the files slice (delete the tmp file for stdin copying) in cache mode.

Comment (Posrabi) [reply main.go:340]: And yeah it could be importFile

Comment (Posrabi) [reply main.go:340]: Yeah if you don't mind, you can merge it.

Comment (eatonphil) [reply main.go:340]: Ok! I'll open a PR for it after this is merged so you can comment on it if you want.

I'm ok to merge this now then.

Comment (eatonphil) [reply main.go:340]: Ready to merge the DataStation PR first?

Comment (Posrabi) [reply main.go:340]: Yeah!

Comment (eatonphil) [reply main.go:340]: Ok, that's merged. Update go.mod in here and then we'll merge this

Comment (Posrabi) [reply main.go:339]: Done.

**Comment by eatonphil** [comment]
Your two new tests are failing in Github Actions, try committing dsq with `--verbose` in your test cases if you don't understand why they're failing.

File: scripts/test.py [hunk 25]
```diff
@@ -213,6 +213,33 @@ def test(name, to_run, want, fail=False, sort=False):
...
+"""
+
+cmd("curl https://s3.amazonaws.com/nyc-tlc/trip+data/yellow_tripdata_2021-04.csv -o taxi.csv")
+test("Caching from file", to_run, want, sort=True)
+
+to_run = """
+cat taxi.csv | ./dsq --cache -s csv "SELECT passenger_count, COUNT(*), AVG(total_amount) FROM {} GROUP BY passenger_count ORDER BY COUNT(*) DESC"
+"""
+
+test("Caching from pipe", to_run, want, sort=True)
```
[the remark this answers is not in the document]
Comment (Posrabi) [reply scripts/test.py:241]: Yeah!

**Comment by eatonphil** [comment]
Any idea why tests are still failing?

**Comment by Posrabi** [comment]
> Any idea why tests are still failing?

I honestly don't. In the previous comment, do you mean putting --verbose in to_run?

**Comment by eatonphil** [comment]
>  In the previous comment, do you mean putting --verbose in to_run?

dsq hides DataStation logs by default.

If you run dsq with `--verbose` so `dsq --verbose blah blah` it will show the DataStation logs too which can be useful for error messages.

But actually it might just be my Python test script that is hiding the error messages.

**Comment by Posrabi** [comment]
> > In the previous comment, do you mean putting --verbose in to_run?
> 
> dsq hides DataStation logs by default.
> 
> If you run dsq with `--verbose` so `dsq --verbose blah blah` it will show the DataStation logs too which can be useful for error messages.
> 
> But actually it might just be my Python test script that is hiding the error messages.

It's cause of "SQLite not found on disk" log. We don't really need it right?

**Comment by eatonphil** [comment]
> It's cause of "SQLite not found on disk" log. We don't really need it right?

This means that caching isn't happening. You're trying to test caching though right?

**Comment by Posrabi** [comment]
> > It's cause of "SQLite not found on disk" log. We don't really need it right?
> 
> This means that caching isn't happening, right? You're trying to test caching though right?

It's for when we are in cache mode and there isn't a sqlite file, there will be a log: "SQLite file is not found on disk, creating a new one..."
Yeah we are still testing for caching.

**Comment by eatonphil** [comment]
btw I don't think piping works on windows. So you can make that test conditional on if it's being run within Windows.

**Comment by eatonphil** [comment]
This is the error I get on Windows:

image

**Comment by eatonphil** [comment]
Nope tried it and it's not that. Pretty sure it's because you're trying to open SQLite when it doesn't exist somewhere.

**Comment by Posrabi** [comment]
does windows have a tmp directory?

**Comment by eatonphil** [comment]
Oh yeah you can't hardcode that to /tmp/ :D . Windows puts it inside of the user's home directory. You could use the TmpFile API in Go and then maybe just rename that file within the same directory? dunno

**Comment by Posrabi** [comment]
image
it works on my windows ...

**Comment by eatonphil** [comment]
Hmm yeah I'm not sure why it's failing. It passes for me too on my machine.

File: .github/workflows/pull_requests.yml [hunk 26]
```diff
@@ -10,55 +10,60 @@ jobs:
...
 
-    - run: ./scripts/ci/prepare_windows.ps1
-      shell: pwsh
-    - run: go build -trimpath -buildmode=pie -mod=readonly -modcacherw
-    - run: ./scripts/test.py
-      shell: bash
+      - run: ./scripts/ci/prepare_windows.ps1
+        shell: pwsh
+      - run: go build -trimpath -buildmode=pie -mod=readonly -modcacherw
+      # if you call os.TempDir() that path would be \blabla\Runner~1\blabla the ~ will break some apps.
```
Comment (eatonphil) [inline .github/workflows/pull_requests.yml:45]: Actually you found a bug in DataStation here! Tildes should be valid anywhere it was my own bad resolution code that broke it. Revert this change and update go.mod to take in https://github.com/multiprocessio/datastation/pull/228 when I merge that.

Comment (Posrabi) [reply .github/workflows/pull_requests.yml:45]: Okay

Comment (eatonphil) [reply .github/workflows/pull_requests.yml:45]: That PR is merged now. Give it a shot! I'm going to the doctor so I can approve/merge in 3 hours or so

Comment (Posrabi) [reply .github/workflows/pull_requests.yml]: it works!

**Comment by eatonphil** [comment]
Great work!
